### PR TITLE
feat(pipelines/gaming_analytics): add Java streaming pipeline for the gaming analytics guide

### PR DESCRIPTION
**Pull request by iht**
## What this adds

The Beam pipeline for the gaming analytics solution guide, in Java. The Terraform module for this guide was merged in #277 and already generated an environment script for a pipeline that did not exist yet; this fills that gap.

`pipelines/gaming_analytics_java/` follows the same Gradle + spotless + errorprone layout as `etl_integration_java` and `clickstream_analytics_java`.

### Pipeline

```
Pub/Sub (gameplay events)
  └─ parse + validate
      └─ enrich with player features from Bigtable
          └─ score with the configured recommender
              ├─ Pub/Sub  → in-game activation
              └─ BigQuery → analytics (Storage Write API)
```

Every failure path — parse, enrichment, scoring, and failed BigQuery inserts — is flattened into the dead-letter topic rather than dropping the element.

Inference sits behind a `Recommender` interface with two implementations: `local` runs the bundled CPU recommender in-process (`gpu` is accepted as an alias, since that is what Terraform exports), and `vertex` calls a Vertex AI endpoint. The recommender is injected into `createPipeline`, so the entire graph runs on the DirectRunner in unit tests with no cloud dependency.

### Security guardrails

* `--usePublicIps=false`
* Dedicated worker service account via `--serviceAccount`
* `--subnetwork`, with a fallback to `NETWORK` for environments generated by older versions of the Terraform code
* GPU accelerators are attached only when `USE_GPU_ACCELERATOR=true`: the bundled recommender is CPU only, so requesting them by default would consume GPU quota for nothing

### Also in this PR

* **Terraform bug fix** (`terraform/gaming_analytics/main.tf`): `ACCELERATOR_OPT` was written unquoted into the generated environment script, so sourcing it failed with `count:1: command not found` and truncated the value to `worker_accelerator=type:nvidia-l4`, losing the count and the driver version. The env file is now also generated into `pipelines/gaming_analytics_java/scripts/`.
* The empty `pipelines/gaming_analytics/` placeholder is folded into the new `_java` directory.
* Docs: added the gaming analytics row (and the previously missing IoT row) to `pipelines/README.md`; corrected the statement in `use_cases/Gaming_Analytics.md` that the guide uses Beam's `Enrichment` and `RunInference` turnkey transforms, which are Python only.

## Testing

Local:

* `./gradlew spotlessCheck build` — passes, **68 unit tests** across 11 classes, including a DirectRunner end-to-end test of the graph.
* `terraform fmt -check` and `terraform validate` — pass.
* `pylint --rcfile ../pylintrc .` on the helper scripts — 10.00/10.
* The `ACCELERATOR_OPT` fix was verified by rendering the `local_file` and sourcing the result before and after.

> [!IMPORTANT]
> **Not yet run in Google Cloud.** The Bigtable `readRows` path, the Vertex AI request/response round trip (exercised only against synthetic JSON) and the GPU path have not been validated against live resources. Please treat this as pending the CONTRIBUTING.md "run in the cloud" requirement.

## Known follow-ups

* Terraform still defaults `inference_mode = "gpu"` with `g2-standard-4`, while the shipped Java recommender is CPU only. Documented in the pipeline README, but the Terraform defaults over-provision for this pipeline.
* Bigtable lookups have no server-side column-family or version filter; they read all families and all cell versions per lookup. This mirrors the existing `clickstream_analytics_java` implementation.

### CODE REVIEW

File: pipelines/gaming_analytics_java/scripts/generate_gameplay_events.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
+"""Generates and publishes synthetic gameplay events to Pub/Sub."""
```
Comment (iht) [inline pipelines/gaming_analytics_java/scripts/generate_gameplay_events.py:14]: Missing copyright header

File: pipelines/gaming_analytics_java/scripts/populate_bigtable.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
+"""Populates the Cloud Bigtable feature store with sample player features."""
```
Comment (iht) [inline pipelines/gaming_analytics_java/scripts/populate_bigtable.py:14]: Missing copyright header

File: pipelines/gaming_analytics_java/src/main/java/com/google/cloud/dataflow/solutions/gaming_analytics/data/GamingObjects.java [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+     */
+    @DefaultSchema(AutoValueSchema.class)
+    @AutoValue
+    public abstract static class EnrichedEvent implements Serializable {
+
+        @SchemaFieldName("event")
+        public abstract GameplayEvent getEvent();
+
+        @SchemaFieldName("features")
+        public abstract Map<String, String> getFeatures();
```
Comment (iht) [inline pipelines/gaming_analytics_java/src/main/java/com/google/cloud/dataflow/solutions/gaming_analytics/data/GamingObjects.java]: This should return a ImmutableMap

File: pipelines/gaming_analytics_java/src/main/java/com/google/cloud/dataflow/solutions/gaming_analytics/load/PubSubPublishers.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+    /** Builds the Pub/Sub message carrying a recommendation. */
+    public static PubsubMessage toPubsubMessage(Recommendation recommendation) {
+        Map<String, String> attributes = new HashMap<>();
+        attributes.put("player_id", recommendation.getPlayerId());
+        if (recommendation.getEventType() != null) {
+            attributes.put("event_type", recommendation.getEventType());
+        }
+        if (recommendation.getRecommendation() != null) {
+            attributes.put("recommendation", recommendation.getRecommendation());
```
Comment (iht) [inline pipelines/gaming_analytics_java/src/main/java/com/google/cloud/dataflow/solutions/gaming_analytics/load/PubSubPublishers.java]: Why are these in the attributes and the payload at the same time?

File: pipelines/gaming_analytics_java/src/main/java/com/google/cloud/dataflow/solutions/gaming_analytics/options/GamingAnalyticsOptions.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    void setOutputTopic(String value);
+
+    @Validation.Required
+    @Description("Pub/Sub dead-letter topic for unprocessable elements (ERROR_TOPIC)")
+    String getErrorTopic();
+
+    void setErrorTopic(String value);
+
+    @Description("Project owning the Bigtable feature store. Defaults to the Dataflow project.")
```
Comment (iht) [inline pipelines/gaming_analytics_java/src/main/java/com/google/cloud/dataflow/solutions/gaming_analytics/options/GamingAnalyticsOptions.java]: This should be optional, and by default the same as Dataflow

File: use_cases/Gaming_Analytics.md [hunk 6]
```diff
@@ -4,7 +4,9 @@ Real-time gaming analytics refers to collecting, processing, and analyzing playe
...
+This reference architecture demonstrates how to ingest gameplay events from the gaming platform servers via **Cloud Pub/Sub**, hydrate them with player features held in **Cloud Bigtable**, score them with a model (either running locally on the workers or deployed behind a **Vertex AI** endpoint), publish recommendations back to a **Pub/Sub** topic for immediate in-game activation, and persist analytical records to **BigQuery**.
+
+In Python, the hydration and scoring steps map directly onto the turnkey `Enrichment` and `RunInference` transforms. The sample pipeline in this repository is written in Java, where those transforms are not available, so it performs the feature lookups with the Cloud Bigtable data client and the scoring behind a pluggable `Recommender` interface. The shape of the pipeline is the same in both languages.
```
Comment (iht) [inline use_cases/Gaming_Analytics.md]: RunInference can be used in Java. Check this documentation about that: https://beam.apache.org/documentation/ml/multi-language-inference/

**Comment by iht** [comment]
Thanks — the RunInference comment was the right catch, and it invalidated a design decision rather than just a sentence of prose. Pushed as b04828c.

### RunInference

You were right, and so was the Runner v2 instinct. The pipeline now uses the real cross-language `org.apache.beam.sdk.extensions.python.transforms.RunInference` (`beam-sdks-java-extensions-python`, same pinned Beam 2.76.0), via `RunInference.ofKVs(MODEL_LOADER, PREDICTION_SCHEMA, RowCoder.of(CONTEXT_SCHEMA))` with the Python `KeyedModelHandler(SklearnModelHandlerNumpy(model_uri))`. The keyed variant carries the player id across the language boundary with the element, so rebuilding the recommendation needs no re-join and the activation path gains no shuffle. `--experiments=use_runner_v2` is now in the launch script.

The hand-rolled `Recommender` interface, `LocalRecommender`, `VertexAiRecommender`, the `vertex` mode and `--modelEndpoint` are all deleted; scoring is a pickled scikit-learn model in the Python SDK harness on the worker, produced by the new `scripts/train_model.py` and passed as `--modelUri`.

Two findings worth flagging:

* **`pandas` has to be listed explicitly.** `apache_beam.ml.inference.sklearn_inference` imports it at module scope, and Beam disables `inferExtraPackagesFromModelHandler()` as soon as you pass your own `withExtraPackages(...)`. Without it the model loader cannot be evaluated on the worker — a runtime failure, invisible at build time. There is now a unit test asserting the list is complete and fully `==`-pinned.
* **The three pins are exactly what the Beam 2.76.0 Python harness image already ships** (`sdks/python/container/py3XX/base_image_requirements.txt`), so installing them is a no-op rather than an upgrade underneath the SDK. `train_model.py` refuses to pickle a model under different scikit-learn/numpy versions. Documented as a thing to re-check whenever `beamVersion` moves.

### The other four comments

1. + 2. Copyright headers added to both scripts (and `train_model.py`).
3. `EnrichedEvent.getFeatures()` returns `ImmutableMap`. Narrowing the getter while `Builder.setFeatures` still takes `Map<String, String>` makes AutoValue emit the `ImmutableMap.copyOf(...)` itself and keeps Beam's `TransformingMap` assignable — no `VerifyError`, and no global `-parameters` flag needed. Covered by a schema-coder round-trip test.
4. Attributes removed from the recommendation messages: `player_id`, `event_type` and `recommendation` were already in the JSON body, no consumer in this guide filters or orders on them, and `recommendation` is model output that must not be duplicated into metadata where it can drift; per-player ordering belongs in a Pub/Sub ordering key, not an attribute. `stage` is kept on the dead-letter topic because it genuinely enables a filtered per-stage triage subscription.

### Testing

`./gradlew build` stays hermetic. Cross-language transforms expand at pipeline-construction time and need pip or Docker, so the unit suite exercises the real `toFeatureVectors()` and `toRecommendations()` halves with an in-process stand-in for the boundary hop only, and the real thing lives in `RecommendationInferenceCrossLanguageIT`, skipped via `Assume` unless enabled:

```bash
python scripts/train_model.py --output_path=/tmp/gaming_recommender.pkl
./gradlew test --tests '*RecommendationInferenceCrossLanguageIT' \
  -DrunCrossLanguageTests=true -DcrossLanguageModelUri=/tmp/gaming_recommender.pkl
```

It has a negative case (a broken model loader must be rejected) so the positive one can't pass vacuously. One unit test also decodes a prediction captured from the real Python SDK with the Java coder, pinning the `NumPy float64 → ARRAY<DOUBLE>` schema contract by bytes rather than by assumption.

Verified locally: `./gradlew --offline clean spotlessCheck build` green, 61 tests, 2 integration tests correctly skipped; `pylint --rcfile ../pylintrc .` 10.00/10. The integration test has been run green against a real expansion service.

### Still not verified — the PR stays a draft

The pipeline has never been run end to end: no Runner v2 submission, no Dataflow job, no Bigtable/Pub/Sub/BigQuery traffic. "The expansion service starts" is strictly weaker than "the pipeline works" — the Java DirectRunner cannot execute `beam:transform:external:v1`, so only Dataflow can prove that path. The Cloud NAT note in the README (private-IP workers still have to pull `apache/beam_python3.x_sdk:2.76.0` from Docker Hub) is reasoning, not observation.

### Follow-up, deliberately not in this PR

`terraform/gaming_analytics` still defaults to `inference_mode = "gpu"` with `g2-standard-4` + L4 and still provisions the `gamingAnalyticsPredictor` Vertex AI custom role. scikit-learn is CPU-only and there is no longer a Vertex path, so neither is needed by the shipped pipeline. I left the `.tf` untouched (that module is already merged) and only added a prose warning to its README — happy to send the cleanup as a separate PR if you'd rather not carry the mismatch.

File: terraform/gaming_analytics/README.md [hunk 7]
```diff
@@ -42,6 +42,21 @@ The guide supports two inference topologies. Both are covered by the `inference_
 > [!NOTE]
 > No Dataflow job and no Vertex AI endpoint are created by Terraform. In `vertex` mode you deploy the endpoint separately and grant it to the worker service account.
 
+> [!IMPORTANT]
```
Comment (iht) [inline terraform/gaming_analytics/README.md]: You should update the Terraform code if necessary, and leave this documentation as starting from scratch. There is no need to keep any kind of backwards compatibility. This is a new contribution, and "incompatible' changes are expected and totally ok at this point.

Comment (iht) [reply terraform/gaming_analytics/README.md]: Agreed, and already done in a85b023 — that warning block is gone, along with the thing it was warning about.

The Terraform module is now CPU-only with no compatibility hedging: `inference_mode` is removed outright (not deprecated), the worker defaults to `n1-standard-2`, `ACCELERATOR_OPT`/`INFERENCE_MODE` no longer appear in the generated environment script, the worker disk drops back to 50 GB, and `aiplatform.googleapis.com`, the `gamingAnalyticsPredictor` custom role and its binding are deleted. The launch script lost its `USE_GPU_ACCELERATOR` escape hatch too. `terraform fmt -check -recursive`, `validate` and `test` (3 passed) are green, and the module docs now read as a description of what the module does, with no mention of what it used to do.

Taking the "starting from scratch" point further than the Terraform: I also went through the pipeline docs for prose that only made sense as a diff against an unreleased earlier version, and rewrote two spots in e306faf — the model was described as being fitted on "the same rules the previous rule-based recommender used", and the DirectRunner limitation was phrased as something that "can no longer" be done rather than a plain property of a cross-language pipeline.

One consequence worth naming explicitly, since you said incompatible changes are fine: an existing `terraform.tfvars` carrying `inference_mode` will now fail with an unsupported-variable error rather than being silently ignored. That seems right for a guide nobody has deployed yet, but it is a hard break rather than a soft one.
